### PR TITLE
[v6r14] Fixes in Staging, DM, FTS

### PR DESCRIPTION
**Pull request by phicharp**
* Don't fail getting  metadata for staging if at least one staged replica found
* Allow to not get URLs in DM 
* Fix printout in optimizer that was saying "single site" and "multiple sites" in two consecutive lines ;-)
* Add recovery of FTS files (in FTSAgent) that can be left in weird statuses when the agent dies

### CODE REVIEW

**Comment by fstagni** [comment]
Review OK

**Comment by phicharp** [comment]
Please re-review... Sorry, found many weird cases of transfers waiting since February with requests and FTSFiles in weird states... Implemented some emergency recoveries

File: WorkloadManagementSystem/Executor/JobScheduling.py [hunk 1]
```diff
@@ -222,16 +222,12 @@ def optimizeJob( self, jid, jobState ):
...
-      for site in banned:
-        if site in filtered:
-          filtered.remove( site )
-    return filtered
+    filtered = set( sites )
+    if active and isinstance( active, ( list, set, dict ) ):
+      filtered &= set( active )
+    if banned and isinstance( banned, ( list, set, dict ) ):
+      filtered -= set( banned )
+    return list( filtered )
```
Comment (fstagni) [inline WorkloadManagementSystem/Executor/JobScheduling.py]: This is not fully correct. In fact, it wasn't correct even before.

I made what is the correct fixes in https://github.com/fstagni/DIRAC/commit/02f90336e6bd2626338fbec2f80309b6ecdfa9a8
where I also added the unit tests.

Comment (phicharp) [inline WorkloadManagementSystem/Executor/JobScheduling.py]: Sorry but I don't understand what is in your commit ;-)
I had understood that 'active' was the sites selected by the user, therefore requires an AND with where the files are. My PR is just a rewrite of what was there before, right?
I don't understand what this is supposed to do:

filtered = filtered - set( active ) | filtered

Comment (fstagni) [inline WorkloadManagementSystem/Executor/JobScheduling.py]: I know you just re-wrote it in a better way, in fact as I said it was not fully correct even before.

My change is done to cover a very special case indeed, that is for example when user requires [SiteA], but SiteA is not in the list of "allowed" nor "removed".  Basically it is for https://github.com/fstagni/DIRAC/blob/02f90336e6bd2626338fbec2f80309b6ecdfa9a8/WorkloadManagementSystem/Executor/test/Test_Executor.py#L70

Comment (fstagni) [inline WorkloadManagementSystem/Executor/JobScheduling.py]: I am not sure what's covered will really ever happen, but it's for completeness.

Comment (phicharp) [inline WorkloadManagementSystem/Executor/JobScheduling.py]: ... I didn't re-write it recently, it is as was originally in the PR...
Still Federico sorry but I don't understand your line of code... For me this is not doing anything, as filtered is unaffected??? You subtract the sites requested by the user (why?) and then you OR it with filtered which results in filtered = filtered, right?

File: DataManagementSystem/Agent/FTSAgent.py [hunk 2]
```diff
@@ -521,6 +521,10 @@ def processRequest( self, request ):
               if finishedFiles:
                 log.warn( "%s is %s while replication was Finished to %s, update" % ( ftsFile.LFN, ftsFile.Status, targetSE ) )
                 ftsFilesDict['toUpdate'] += finishedFiles
+            # identify Active transfers for which there is no FTS job any longer and reschedule them
+            for ftsFile in [f for f in ftsFiles if f.Status == 'Active' and f.TargetSE in missingReplicas.get( f.LFN, [] )]:
+              if not [ftsJob for ftsJob in ftsJobs if ftsJob.FTSGUID == ftsFile.FTSGUID]:
```
Comment (chaen) [inline DataManagementSystem/Agent/FTSAgent.py]: Maybe using the 'filter' function would be more self explanatory ?

File: DataManagementSystem/Agent/FTSAgent.py [hunk 3]
```diff
@@ -531,6 +535,20 @@ def processRequest( self, request ):
               if not regOp:
                 ftsFilesDict['toReschedule'].append( ftsFile )
 
+            # Recover files that are Failed but were not spotted
+            for ftsFile in [f for f in ftsFiles if f.Status == 'Failed' and f.TargetSE in missingReplicas.get( f.LFN, [] )]:
```
Comment (chaen) [inline DataManagementSystem/Agent/FTSAgent.py]: filter function as well ?

File: WorkloadManagementSystem/Executor/JobScheduling.py (hunk 1 above)
Comment (fstagni) [inline WorkloadManagementSystem/Executor/JobScheduling.py]: mmm... yes, indeed... in fact I don't see why we should bother about active sites while the only thing to exclude IMHO are banned ones.

**Comment by phicharp** [comment]
Closed as superseded by another PR

**Pull request closed without merging**